Makefile

```diff
@@ -18,6 +18,52 @@ $(MAKEFILE):
 GOTEST_INTEGRATION = $(GOTEST) -parallel 1 -count 1 -tags=integration -ldflags "$(LD_FLAGS)"
 
 OS := $(shell uname)
+CURRENT_BUILD_PATH := $(BUILD_PATH)
+PREV_BUILD_PATH := build-prev
+NEXT_BUILD_PATH := build-next
+
+CURRENT_VERSION := $(shell git rev-parse HEAD)
+
+# Latest non release candidate version
+PREV_VERSION ?= $(shell git describe --abbrev=0 --tags --exclude "*-rc*")
+# Current commit
+NEXT_VERSION ?= $(CURRENT_VERSION)
+
+# Targets for integration testing
+integration-clean-previous:
+	rm -rf $(PREV_BUILD_PATH)
+	$(GOCLEAN) .
+integration-prepare-build-previous:
+	git checkout -q $(PREV_VERSION)
+	$(eval BUILD_PATH = $(PREV_BUILD_PATH))
+integration-build-previous: integration-clean-previous integration-prepare-build-previous build docker-build
+	git checkout -q $(CURRENT_VERSION)
+	echo "Built version $(PREV_VERSION)"
+
+integration-clean-next:
+	rm -rf $(NEXT_BUILD_PATH)
+	$(GOCLEAN) .
+integration-prepare-build-next:
+	git checkout -q $(NEXT_VERSION)
+	$(eval BUILD_PATH = $(NEXT_BUILD_PATH))
+integration-build-next: integration-clean-next integration-prepare-build-next build docker-build
+	git checkout -q $(CURRENT_VERSION)
+	echo "Built version $(NEXT_VERSION)"
+
+integration-clean-current:
+	rm -rf $(CURRENT_BUILD_PATH)
+	$(GOCLEAN) .
+integration-prepare-build-current:
+	git checkout -q $(CURRENT_VERSION)
+	$(eval BUILD_PATH = $(CURRENT_BUILD_PATH))
+integration-build-current: integration-clean-current integration-prepare-build-current build docker-build
+	git checkout -q $(CURRENT_VERSION)
+	echo "Built version $(CURRENT_VERSION)"
+
+integration-build-all:
```

Comment:
have you thought about using pre-built releases from github instead of building them locally?
I'm not sure what is better.
Current approach is more flexible. But using pre-built releases, in theory, are better reproducible. (something can change over time and the local build won't be exact the same as the one from github)

Reply:
> Current approach is more flexible.

Actually, I think that it doesn't have much sense to test a prev version that is not an actual release available on GH (I don't think that we're going to test a specific unreleased version as `PREV`).  And as you said we're actually interested in the release available on GH. Even if the build artifact should be exactly the same, downloading it should also speed-up testing and aslo won't require building the docker image for the daemon.

Reply:
Given that this PR is already big, do you mind approaching this separately?